File: golang/pkg/chip8/memory.go
package chip8

type Memory struct {
	ram		[4096]byte
}

func (c *Memory) LoadData(addr uint16, data []byte) {
	for i := 0; i < len(data); i++ {
		c.ram[int(addr) + i] = data[i]
	} 
}

func (c *Memory) get(addr uint16) byte {
	return c.ram[addr]
}

func (c *Memory) set(addr uint16, value byte) {
	c.ram[addr] = value
}

File: rust/src/chip8/memory.rs
use super::{RAM_SIZE, SPRITE_CHARS, SPRITE_CHARS_ADDR};

pub struct Memory {
    ram: [u8; RAM_SIZE], // RAM
}

impl Memory {
    pub fn new() -> Self {
        let mut memory = Memory::default();

        // load predefined sprites
        for (i, sprite) in SPRITE_CHARS.iter().enumerate() {
            let p = SPRITE_CHARS_ADDR as usize + i * sprite.len();
            memory.load_data(p, sprite);
        }
        memory
    }

    pub fn load_data(&mut self, addr: usize, data: &[u8]) {
        self.ram[addr..addr + data.len()].copy_from_slice(data);
    }

    pub fn get(&self, addr: usize) -> u8 {
        self.ram[addr]
    }

    pub fn set(&mut self, addr: usize, value: u8) {
        self.ram[addr] = value;
    }
}

impl Default for Memory {
    fn default() -> Self {
        Self { ram: [0; 4096] }
    }
}
